submissions/angels_unit_tests/angels_unit_tests.md

```diff
@@ -0,0 +1,55 @@
+
+I am confident now saying that we do not make coding mistakes anymore, code slipping through the cracks, which would result in a hard crash of the mods. Nowadays the issues are likely edge cases, unhandled cases or a strange modpack configuration. Most frequently we receive bugs from people who play with a few Angel's mods and a few Bob mods. The reason for this is simple. There are many configurations with many settings, which result in many oversights. So we began creating a list of configurations we had to test. [*pezzawinkle*](https://mods.factorio.com/user/pezzawinkle) recently made [a not-so-small list](https://github.com/Arch666Angel/mods/blob/master/Config_Testing_Regime.txt) with all configurations he tests every time he prepairs a release.
+
+Personally, I test every change I make with all the configurations I impacted. The most tedious part of fixing bugs is having to reload all Angel's mods into the mod folder over and over again. On the 11th of April 2020 I created [a small python script](https://github.com/Arch666Angel/mods/pull/188/commits/e37f2673f234e0c9271f56ecc6d4934b0d83cd50) to update all Angel's mods to the latest changes.
+
+This script saved so many time, allowing for a sporadic coffee break while it reloaded the mods after tweaking that one line that still didn't behave as intended. By using this script over time, I had another issue, Bob. Yes, Bobbingabout released an update of his mods, and it broke all the workd I just did fixing a bug... On the 3rd of May 2021, almost a year later, I created [another small python script](https://github.com/Arch666Angel/mods/pull/591/commits/3d0a7de0a68c3ec6566cbd45c3599b036ab775bb) to download all bob mods from the portal, rather than doing this ingame one by one.
+
+## The ball sped up rolling downhill
+
+At the end of 2020, we decided to work on a big feature inside Angel's Industries: a decent implementation of the component and technology overhaul. This was the last feature Arch666Angel was working on before he pulled back from modding. On the 26th of May 2021 (half a year since the last release), we presented the new version of the overhauls to the public, after which the bug reports started rolling in quickly.
+
+We realised that with the overhauls we created many more configuration, with even more settings. At this point it becomes quite though to maintain and test this completely by hand. I started thinking about creating some [unit tests](https://en.wikipedia.org/wiki/Unit_testing) to take this burden of our shoulders. After some discussions and ideas from the great modding community I finally decided on the 30th of June 2021 to commit myself at creating the unit test infrastructure, groupted in a separate mod: [Angel's Dev - Unit Tests](https://github.com/Arch666Angel/mods/pull/634/commits/d59328e81235e26beb88ba66659cb78315827a00). A small note, You will not find this mod on the modportal.
```

Comment:
it becomes quite `tough`

Reply:
Fixed in https://github.com/AlternativeFFFF/Alt-F4/pull/551/commits/0d2bc1d8be83f76d557f22d7d39ba53b96e0aa9f